File: Week2/Queues/src/RandomizedQueue.java
import java.util.Iterator;
import java.util.NoSuchElementException;


public class RandomizedQueue<Item> implements Iterable<Item> {

    private Item[] queue;
    private int size;
    private int capacity;
    
    public RandomizedQueue() {
        capacity = 2;
        resizeQueue();
    }
    
    private void resizeQueue() {
        Item[] tmp = (Item[]) new Object[capacity];
        
        for (int i = 0; i < size; i++) {
            tmp[i] = queue[i];
        }
        
        queue = tmp;
    }
    
    private void checkIfQueueIsEmpty() {
        if (size() == 0)
            throw new NoSuchElementException();
    }
    
    public boolean isEmpty() {
        return size() == 0;
    }
    
    public int size() {
        return size;
    }
    
    public void enqueue(Item item) {
        if (item == null)
            throw new NullPointerException();
        
        if (size == capacity) {
            capacity *= 2;
            resizeQueue();
        }
        
        queue[size] = item;
        size++;
    }
    
    public Item dequeue() {
        
        checkIfQueueIsEmpty();
        
        int randomIndex = StdRandom.uniform(size);
        Item randomItem = queue[randomIndex];
        size--;
        queue[randomIndex] = queue[size];
        queue[size] = null;
        
        if (size > 4 && size < capacity / 4) {
            capacity /= 2;
            resizeQueue();
        }
        
        return randomItem;
    }
    
    public Item sample() {
        
        checkIfQueueIsEmpty();
        
        int randomIndex = StdRandom.uniform(size);
        Item randomItem = queue[randomIndex];
        return randomItem;
    }
    
    private class RandomizedQueueIterator implements Iterator<Item> {        
        
        private Item[] iteratorQueue;
        private int current;
        
        public RandomizedQueueIterator() {
            current = 0;
            
            iteratorQueue = (Item[]) new Object[size];
            for (int i = 0; i < size; i++) {
                iteratorQueue[i] = queue[i];
            }
            
            StdRandom.shuffle(iteratorQueue);
        }
        
        @Override
        public boolean hasNext() {
            return current < size;
        }

        @Override
        public Item next() {
            if (current >= size() || isEmpty())
                throw new NoSuchElementException();
            return iteratorQueue[current++];
        }

        @Override
        public void remove() {
            throw new UnsupportedOperationException();
        }
        
    }
    
    @Override
    public Iterator<Item> iterator() {
        return new RandomizedQueueIterator();
    }
    
}

File: Week2/Subset/Subset/RandomizedQueue.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Subset
{
    public class RandomizedQueue<T> : IEnumerable<T>
    {
        private T[] _queue;
        private int _size;
        private int _capacity;

        public RandomizedQueue()
        {
            _capacity = 4;
            _queue = new T[_capacity];
        }

        public bool IsEmpty()
        {
            return Size() == 0;
        }

        public int Size()
        {
            return _size;
        }

        public void Enqueue(T item)
        {
            if (_size == _capacity)
            {
                _capacity *= 2;
                var tmp = new T[_capacity];
                for (int i = 0; i < _size; i++)
                {
                    tmp[i] = _queue[i];
                }

                _queue = tmp;
            }

            _queue[_size] = item;
            _size++;
        }

        public T Dequeue()
        {
            var randomIndex = new Random().Next(_size);
            var randomItem = _queue[randomIndex];
            _size--;
            _queue[randomIndex] = _queue[_size];
            _queue[_size] = default(T);

            if (_size <= _capacity / 4)
            {
                _capacity /= 2;
                var tmp = new T[_capacity];
                for (int i = 0; i < _size; i++)
                {
                    tmp[i] = _queue[i];
                }

                _queue = tmp;
            }

            return randomItem;
        }

        public T Sample()
        {
            var randomIndex = new Random().Next(_size);
            var randomItem = _queue[randomIndex];
            return randomItem;
        }

        #region IEnumerable
        public IEnumerator<T> GetEnumerator()
        {
            for (int i = _size - 1; i >= 0; i--)
            {
                yield return _queue[i];
            }
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        #endregion
    }
}